# SKLEARN 16 | Random Forest Classification | Hutan Acak | Belajar Machine Learning Dasar

Materi Pembelajaran dikutip dari sumber: https://www.youtube.com/watch?v=yKovaQ6tyV8

Referensi: https://en.wikipedia.org/wiki/Random_forest

## General ML Model Training

![](./images/general_ML_model_training.png)

## Ensemble Learning: heterogeneous & homogeneous

Referensi: https://en.wikipedia.org/wiki/Ensemble_learning

![](./images/ensemble_learning.png)

+ Ensemble Learning adalah teknik menggabungkan beberapa model untuk prediksi.
+ Training set yang tersedia akan digunakan untuk mentraining beberapa model.
+ Terdapat 3 model yaitu KNN, SVM, dan Decision Tree.
+ y_pred 1, y_pred 2, y_pred 3 akan disatukan menggunakan nilai Mean atau Mode. Lalu hasilnya akan dipresentasikan menjadi y_pred final.
+ Jika kita memadukan beberapa machine learning yang sejenis dapat disebut **homogeneous**, sedangkan beberapa machine learning yang berbeda dapat disebut **heterogeneous**.

## Bagging: Bootstrap Aggregating

Referensi: https://en.wikipedia.org/wiki/Bootstrap_aggregating

![](./images/bagging.png)

+ Berikut kita mempelajari Homogeneous Ensemble Learning.
+ Bootstrap Aggregating menerapkan Random Sampling with Replacement terhadap training set yang dimiliki dan menghasilkan beberapa training set baru sejumlah model yang kita training.
+ Training set baru yang dihasilkan dikenal dengan istilah Bag. Terdapat 3 buah bag yaitu Bag 1, Bag 2 dan Bag 3.
+ Bag akan digunakan untuk mentraining machine learning yang sejenis.
+ Trained Model 1, Trained Model 2, dan Trained Model 3 akan digunakan untuk melakukan prediksi, 3 hasil prediksi akan disatukan dengan Majority Voting yaitu Mean or Mode.

## Random Forest

Referensi: https://en.wikipedia.org/wiki/Random_forest

![](./images/random_forest.png)

+ Random Forest merupakan implementasi dari Homogeneous Ensemble Learning yang menerapkan Decision Tree.
+ Terdapat model yang sejenis yaitu Decision Tree.
+ Random Forest terbentuk dari sekumpulan Decision Tree atau pohon keputusan.
+ Random Forest menerapkan Features Randomness, dimana setiap bag yang dihasilkan akan mengadopsi sejumlah features yang dipilih secara acak dari training set sumbernya.
+ Terdapat trained model yang beragam, walaupun menggunakan machine learning yang sejenis.
+ Prediksi yang dihasilkan akan disatukan melalui proses Majority Voting untuk y_pred final.

## Dataset

Referensi: https://en.wikipedia.org/wiki/Iris_flower_data_set

In [1]:
from sklearn.datasets import load_iris

X, y = load_iris(return_X_y=True)

print(f'Dimensi Feature: {X.shape}')
print(f'Class: {set(y)}')

Dimensi Feature: (150, 4)
Class: {0, 1, 2}


Berikut terdapat 150 baris dan 4 kolom, lalu terdapat 3 class.

In [2]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.3,
                                                    random_state=0)

## Classification dengan RandomForestClassifier

In [3]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators=100, 
                               random_state=0)

model.fit(X_train, y_train)

RandomForestClassifier(random_state=0)

n_estimators=100 yaitu berapa banyak decision tree yang mau ditraining.

## Evaluasi Model

In [4]:
from sklearn.metrics import classification_report

y_pred = model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       1.00      0.94      0.97        18
           2       0.92      1.00      0.96        11

    accuracy                           0.98        45
   macro avg       0.97      0.98      0.98        45
weighted avg       0.98      0.98      0.98        45

